crimson/os/seastore: LBACursor based LBAManager/TM interfaces and the related ObjectDataHandler refactor

xxhdx1985126 opened the pull request:
Signed-off-by: Xuehan Xu <xuxuehan@qianxin.com>


<!--
  - Please give your pull request a title like

      [component]: [short description]

  - Please use this format for each git commit message:

      [component]: [short description]

      [A longer multiline description]

      Fixes: [ticket URL on tracker.ceph.com, create one if necessary]
      Signed-off-by: [Your Name] <[your email]>

    For examples, use "git log".
-->

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "quincy"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

- When filling out the below checklist, you may click boxes directly in the GitHub web UI.  When entering or editing the entire PR message in the GitHub web UI editor, you may also select a checklist item by adding an `x` between the brackets: `[x]`.  Spaces and capitalization matter when checking off items this way.

## Checklist
- Tracker (select at least one)
  - [ ] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [x] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [x] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [x] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [x] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins test classic perf` [Jenkins Job](https://jenkins.ceph.com/view/all/job/ceph-perf-classic/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-perf-pull-requests/config/definitions/ceph-perf-pull-requests.yml)
- `jenkins test crimson perf` [Jenkins Job](https://jenkins.ceph.com/view/all/job/ceph-perf-crimson/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-perf-pull-requests/config/definitions/ceph-perf-pull-requests.yml)
- `jenkins test signed` [Jenkins Job](https://jenkins.ceph.com/job/ceph-pr-commits/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-pr-commits/config/definitions/ceph-pr-commits.yml)
- `jenkins test make check` [Jenkins Job](https://jenkins.ceph.com/job/ceph-pull-requests/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-pull-requests/config/definitions/ceph-pull-requests.yml)
- `jenkins test make check arm64` [Jenkins Job](https://jenkins.ceph.com/job/ceph-pull-requests-arm64/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-pull-requests-arm64/config/definitions/ceph-pull-requests-arm64.yml)
- `jenkins test submodules` [Jenkins Job](https://jenkins.ceph.com/view/all/job/ceph-pr-submodules/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-pr-submodules/config/definitions/ceph-pr-commits.yml)
- `jenkins test dashboard` [Jenkins Job](https://jenkins.ceph.com/view/all/job/ceph-dashboard-pull-requests/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-dashboard-pull-requests/config/definitions/ceph-dashboard-pull-requests.yml)
- `jenkins test dashboard cephadm` [Jenkins Job](https://jenkins.ceph.com/view/all/job/ceph-dashboard-cephadm-e2e/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-dashboard-cephadm-e2e/config/definitions/ceph-dashboard-cephadm-e2e.yml)
- `jenkins test api` [Jenkins Job](https://jenkins.ceph.com/view/all/job/ceph-api/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-pr-api/config/definitions/ceph-pr-api.yml)
- `jenkins test docs` [ReadTheDocs](https://readthedocs.org/projects/ceph/) | [Github Workflow Definition](https://github.com/ceph/ceph/blob/main/.readthedocs.yml)
- `jenkins test ceph-volume all` [Jenkins Jobs](https://jenkins.ceph.com/view/ceph-volume%20PR/) | [Jenkins Jobs Definition](https://github.com/ceph/ceph-build/blob/main/ceph-volume-cephadm-prs/config/definitions/ceph-volume-pr.yml)
- `jenkins test windows` [Jenkins Job](https://jenkins.ceph.com/job/ceph-windows-pull-requests/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-windows-pull-requests/config/definitions/ceph-windows-pull-requests.yml)
- `jenkins test rook e2e` [Jenkins Job](https://jenkins.ceph.com/view/all/job/ceph-orchestrator-rook-e2e/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-rook-e2e/config/definitions/ceph-orchestrator-rook-e2e.yml)
</details>

xxhdx1985126 commented:
jenkins test make check

xxhdx1985126 commented:
jenkins test make check arm64

anthonyeleven commented on src/common/options/crimson.yaml.in:
```diff
@@ -152,13 +152,6 @@ options:
-  desc: split extent if ratio of total extent size to write size exceeds this value
-  default: 1.25
-# TODO: seastore_obj_data_write_amplification is no longer correct if 
-#       seastore_data_delta_based_overwrite is enabled. So, this should be reconsidered.
```
This is no longer used?

xxhdx1985126 replied:
Yeah, it's very inaccurate.

xxhdx1985126 commented:
@cyx1231st This PR is good for review, please take a look, thanks:-)

xxhdx1985126 commented:
jenkins test make check

xxhdx1985126 commented:
jenkins test windows

xxhdx1985126 commented:
@cyx1231st I've just modified the PR as discussed offline, putting `punch_hole` and related functions in TransactionManager, please take a look, thanks:-)

cyx1231st reviewed (commented):
Sorry I might be slow as a start, need to recall/look at more places to understand the impacts comprehensively.

cyx1231st commented:
^ IIUC this is related to performance optimization because we don't need to lookup the entire lba tree in more places, probably worth to share some rough performance numbers (before and after this PR) if possible.

xxhdx1985126 commented:
> ^ IIUC this is related to performance optimization because we don't need to lookup the entire lba tree in more places, probably worth to share some rough performance numbers (before and after this PR) if possible.

Yeah. We did the perf test in the following way:

1. A single OSD cluster was used, the OSD ran on a single cpu core;
2. The rbd image's was 10GB;
3. We filled the rbd image with 4MB seq writes;
4. Then we did 4KB uniformal random writes to the image;

Here's the result:
![trans-iterator](https://github.com/user-attachments/assets/a42bec0b-23bf-4a3d-af0b-704d5f61f047)

*main branch*:
```
randwrite-4k: (g=0): rw=randwrite, bs=(R) 4096B-4096B, (W) 4096B-4096B, (T) 4096B-4096B, ioengine=rbd, iodepth=64
fio-3.33-61-g4483
Starting 1 process
Jobs: 1 (f=1): [w(1)][100.0%][w=11.9MiB/s][w=3045 IOPS][eta 00m:00s]
randwrite-4k: (groupid=0, jobs=1): err= 0: pid=1815350: Thu Jun  5 17:02:25 2025
  write: IOPS=3040, BW=11.9MiB/s (12.5MB/s)(13.9GiB/1200020msec); 0 zone resets
    slat (nsec): min=1363, max=1375.3k, avg=8247.82, stdev=3847.93
    clat (usec): min=886, max=377044, avg=21042.35, stdev=30019.66
     lat (usec): min=895, max=377050, avg=21050.60, stdev=30019.65
    clat percentiles (msec):
     |  1.00th=[    3],  5.00th=[    3], 10.00th=[    4], 20.00th=[    5],
     | 30.00th=[    6], 40.00th=[    7], 50.00th=[    9], 60.00th=[   11],
     | 70.00th=[   15], 80.00th=[   23], 90.00th=[   81], 95.00th=[   95],
     | 99.00th=[  115], 99.50th=[  140], 99.90th=[  176], 99.95th=[  184],
     | 99.99th=[  199]
   bw (  KiB/s): min= 7024, max=14920, per=100.00%, avg=12165.72, stdev=399.98, samples=2399
   iops        : min= 1756, max= 3730, avg=3041.39, stdev=100.01, samples=2399
  lat (usec)   : 1000=0.01%
  lat (msec)   : 2=0.77%, 4=19.05%, 10=36.75%, 20=20.96%, 50=8.27%
  lat (msec)   : 100=11.04%, 250=3.16%, 500=0.01%
  cpu          : usr=3.22%, sys=1.56%, ctx=1882103, majf=0, minf=14029
  IO depths    : 1=0.1%, 2=0.1%, 4=0.1%, 8=0.1%, 16=0.1%, 32=0.1%, >=64=100.0%
     submit    : 0=0.0%, 4=100.0%, 8=0.0%, 16=0.0%, 32=0.0%, 64=0.0%, >=64=0.0%
     complete  : 0=0.0%, 4=100.0%, 8=0.0%, 16=0.0%, 32=0.0%, 64=0.1%, >=64=0.0%
     issued rwts: total=0,3648202,0,0 short=0,0,0,0 dropped=0,0,0,0
     latency   : target=0, window=0, percentile=100.00%, depth=64

Run status group 0 (all jobs):
  WRITE: bw=11.9MiB/s (12.5MB/s), 11.9MiB/s-11.9MiB/s (12.5MB/s-12.5MB/s), io=13.9GiB (14.9GB), run=1200020-1200020msec

Disk stats (read/write):
  nvme1n1: ios=25495/11218278, merge=5/8749694, ticks=2453/175010, in_queue=177463, util=100.00%
```

*This PR*:
```
randwrite-4k: (g=0): rw=randwrite, bs=(R) 4096B-4096B, (W) 4096B-4096B, (T) 4096B-4096B, ioengine=rbd, iodepth=64
fio-3.37-124-gbcd4
Starting 1 process
Jobs: 1 (f=1): [w(1)][100.0%][w=12.8MiB/s][w=3274 IOPS][eta 00m:00s]
randwrite-4k: (groupid=0, jobs=1): err= 0: pid=440452: Thu Jun  5 16:56:46 2025
  write: IOPS=3263, BW=12.7MiB/s (13.4MB/s)(14.9GiB/1200015msec); 0 zone resets
    slat (nsec): min=1644, max=464720, avg=5412.84, stdev=1779.55
    clat (usec): min=727, max=942182, avg=19605.60, stdev=17946.55
     lat (usec): min=733, max=942186, avg=19611.01, stdev=17946.54
    clat percentiles (msec):
     |  1.00th=[    3],  5.00th=[    4], 10.00th=[    4], 20.00th=[    6],
     | 30.00th=[    8], 40.00th=[   11], 50.00th=[   14], 60.00th=[   18],
     | 70.00th=[   24], 80.00th=[   31], 90.00th=[   44], 95.00th=[   57],
     | 99.00th=[   80], 99.50th=[   89], 99.90th=[  117], 99.95th=[  131],
     | 99.99th=[  155]
   bw (  KiB/s): min= 3296, max=18168, per=100.00%, avg=13065.33, stdev=506.75, samples=2398
   iops        : min=  824, max= 4542, avg=3266.25, stdev=126.67, samples=2398
  lat (usec)   : 750=0.01%, 1000=0.01%
  lat (msec)   : 2=0.17%, 4=10.42%, 10=27.46%, 20=26.54%, 50=28.19%
  lat (msec)   : 100=7.00%, 250=0.22%, 1000=0.01%
  cpu          : usr=2.52%, sys=1.50%, ctx=1768644, majf=0, minf=3332
  IO depths    : 1=0.1%, 2=0.1%, 4=0.1%, 8=0.1%, 16=0.1%, 32=0.1%, >=64=100.0%
     submit    : 0=0.0%, 4=100.0%, 8=0.0%, 16=0.0%, 32=0.0%, 64=0.0%, >=64=0.0%
     complete  : 0=0.0%, 4=100.0%, 8=0.0%, 16=0.0%, 32=0.0%, 64=0.1%, >=64=0.0%
     issued rwts: total=0,3915991,0,0 short=0,0,0,0 dropped=0,0,0,0
     latency   : target=0, window=0, percentile=100.00%, depth=64

Run status group 0 (all jobs):
  WRITE: bw=12.7MiB/s (13.4MB/s), 12.7MiB/s-12.7MiB/s (13.4MB/s-13.4MB/s), io=14.9GiB (16.0GB), run=1200015-1200015msec

Disk stats (read/write):
  nvme0n1: ios=0/11994, sectors=0/288384, merge=0/19755, ticks=0/773, in_queue=773, util=0.56%
```

xxhdx1985126 commented on src/crimson/os/seastore/btree/fixed_kv_btree.h:
```diff
@@ -170,6 +200,67 @@ class FixedKVBtree {
+        }
+
+        assert(parent->get_node_meta().is_parent_of(child_meta));
+        assert(parent->is_viewable_by_trans(c.trans).first);
```
Yeah, will do

xxhdx1985126 commented:
> The following tests FAILED:
	152 - unittest_fault_injector (Timeout)

xxhdx1985126 commented:
jenkins test make check

xxhdx1985126 commented:
jenkins test make check arm64

cyx1231st commented on src/crimson/os/seastore/btree/fixed_kv_btree.h:
```diff
@@ -1749,124 +1933,145 @@ class FixedKVBtree {
+            return iterator::ensure_internal_iertr::template make_ready_future<
+              seastar::stop_iteration>(seastar::stop_iteration::yes);
+          }
+          return iter.ensure_internal(c, split_from + 1
```
Roughly looked at the diff here, can we make sure `ensure_internal()` is called at the beginning with `check_split()`, so that the following logic can be mostly unchanged and keeps to be synchronous (which is a good property), because this is top-down ?

If the above is possible, we can split out 1 cleanup commits to be helpful:
1. a cleanup-only preparation: make `check_split()` interface asynchronous, and only moving the code.
2. apply real changes after, with smaller diff to look at.

cyx1231st reviewed (commented):
looking ..

xxhdx1985126 commented on src/crimson/os/seastore/btree/fixed_kv_btree.h:
```diff
@@ -828,7 +973,21 @@ class FixedKVBtree {
+              return ret.prev(c
+              ).si_then([c](auto it) {
+                return it.next(c);
+              });
```
Because, after removal, we may end up at the end of an leaf node but not the end of btree. In this case, we need to go `prev` and then go `next` to go to the next leaf node.

xxhdx1985126 replied:
Ah, yes.

xxhdx1985126 commented on src/crimson/os/seastore/linked_tree_node.h:
```diff
@@ -1014,6 +1043,18 @@ class ChildNode : public BaseChildNode<ParentT, key_t> {
+    ExtentTransViewRetriever &etvr,
+    key_t key)
+  {
+    return etvr.get_extent_viewable_by_trans<ParentT>(
```
Ah, I think you mean that `find_pending_version` also search `mutation_pending_extents` as in `get_extent_viewable_by_trans`, so we don't need to call the latter. Am I right?

cyx1231st replied:
Per discussion, there are differences between top-down and bottom-up lookups, probably good to split this interface.

xxhdx1985126 commented on src/crimson/os/seastore/linked_tree_node.h:
```diff
@@ -302,27 +302,14 @@ class ParentNode {
-    }
-    ceph_assert((*it)->get_begin() <= key && key < (*it)->get_end());
-    return *it;
+    return {true, find_pending_version(t, key)};
```
Yeah, will do

cyx1231st commented on src/crimson/os/seastore/btree/btree_types.h:
```diff
@@ -279,16 +280,20 @@ struct LBACursor : BtreeCursor<laddr_t, lba::lba_map_val_t> {
+  bool contains(laddr_t laddr) const {
+    return get_laddr() <= laddr && get_laddr() + get_length() > laddr;
+  }
+  boost::intrusive_ptr<LBACursor> duplicate() const {
```
As discussed before, since `LBACursor` is changed to a `boost::intrusive_ptr` in this PR, the ptr itself can be copied, seems to me there is no need to copy-construct the `LBACursor` any more, right?

xxhdx1985126 commented on src/crimson/os/seastore/object_data_handler.cc:
```diff
@@ -1450,26 +556,79 @@ ObjectDataHandler::write_ret ObjectDataHandler::write(
+	[&params, ctx, &data](LBAMapping &mapping, bool merge_left) {
+	  return on_merge(ctx, params, data, mapping, merge_left);
+	}
+      ).si_then([&params, ctx, &data](auto mapping) {
```
I think this is indeed clearer, I'll give it a try:-)

xxhdx1985126 replied:
I moved data read related codes back to `ObjectDataHandler`, and added the following interface to `TM`:
1. TM::cut_mapping(): remove the left/right part the of the mapping;
2. TM::punch_in_hole_mappings(): remove mappings within the provided range;
3. TM::punch_in_mapping_hole(): punch a hole in the provided mapping.

cyx1231st commented:
^ Add more eyes because this PR basically reworks object-data-handler, especially the overwrite part (looks to me it becomes simpler and more efficient, though, refering to the above comments).

xxhdx1985126 commented on src/crimson/os/seastore/object_data_handler.cc:
```diff
@@ -1055,7 +87,8 @@ auto with_objects_data(
 }
 
-ObjectDataHandler::write_ret ObjectDataHandler::prepare_data_reservation(
+ObjectDataHandler::write_iertr::future<LBAMapping>
```
It seems that `std::optional` has such a feature: `std::move(opt)` has the same effect as `std::move(*opt)`. So even if we turn to use `std::optional`, we still need to use `is_null()` outside LBAMapping. So I think maybe we should keep the current way. What do you think?

xxhdx1985126 commented on src/crimson/os/seastore/object_data_handler.cc:
```diff
@@ -1450,26 +556,79 @@ ObjectDataHandler::write_ret ObjectDataHandler::write(
+  ceph_assert(!object_data.is_null());
+  ceph_assert(size <= object_data.get_reserved_data_len());
+  auto data_base = object_data.get_reserved_data_base();
+  auto key = (data_base + size).get_aligned_laddr();
```
It seems that, at present, `laddr_t` is in terms of `laddr_t::UNIT_SIZE` which is `4KB` and is not trivial to make it aligned to other sizes.

On the other hand, `TM::get_block_size()` is also 4KB in most cases. So I think maybe we can later submit a dedicated PR that unifies `laddr_t::UNIT_SIZE` and `TM::get_block_size()` and leave this PR as it is, what do you think? @cyx1231st

xxhdx1985126 commented on src/crimson/os/seastore/object_data_handler.cc:
```diff
@@ -1786,45 +882,52 @@ ObjectDataHandler::clone_ret ObjectDataHandler::clone(
-	ctx.t,
-	d_object_data.get_reserved_data_base(),
-	d_object_data.get_reserved_data_len());
+    return ctx.tm.get_pin(ctx.t, object_data.get_reserved_data_base()
```
I'm adding this interface in #57735 , will push it soon:-)

xxhdx1985126 commented on src/crimson/os/seastore/object_data_handler.cc:
```diff
@@ -1083,307 +116,379 @@ ObjectDataHandler::write_ret ObjectDataHandler::prepare_data_reservation(
   context_t ctx,
   laddr_t data_base,
   objaddr_t offset,
   extent_len_t len,
```
Will submit a dedicated PR for this:-)

cyx1231st reviewed (commented):
Minor suggestions in/below fixed-kv-btree.

Plan to take a look at object-data-handler.

xxhdx1985126 commented on src/crimson/os/seastore/cache.h:
```diff
@@ -462,6 +462,81 @@ class Cache : public ExtentTransViewRetriever {
+	  ++access_stats.cache_lru;
+	  ++stats.access.cache_lru;
+	}
+	if (ret.is_paddr_known) {
```
Um, I tried to remove `needs_step_2` in `get_extent_viewable_by_trans`, but it turned out that it's still needed as omap tree can hit `initial_pending` nodes in top-down searches

cyx1231st commented on src/crimson/os/seastore/object_data_handler.cc:
```diff
@@ -1083,308 +116,634 @@ ObjectDataHandler::write_ret ObjectDataHandler::prepare_data_reservation(
-  auto right = tailbl ?
-    (offset + len + tailbl->length()) :
-    (offset + len);
+struct overwrite_params_t {
```
Pass it among many methods, and arbitrarily modify some members in the middle is not easy to follow.

Ideally, for such struct with complex states, maintain and update only in high-level logic, and isolate it from self-contained lower-level methods with clear boundary.

cyx1231st commented on src/crimson/os/seastore/transaction_manager.h:
```diff
@@ -108,7 +111,35 @@ class TransactionManager : public ExtentCallbackInterface {
+   */
+  get_pin_ret get_containing_pin(
+    Transaction &t,
+    laddr_offset_t laddr) {
```
After a second thought, let's deal with (encapsulate) alignments only inside the ObjectDataHandler -- e.g. use the aligned type laddr_t at the TM boundary.

cyx1231st commented on src/crimson/os/seastore/transaction_manager.h:
```diff
@@ -635,6 +544,27 @@ class TransactionManager : public ExtentCallbackInterface {
 
+  reserve_extent_ret reserve_region(
+    Transaction &t,
+    LBAMapping mapping,
```
For inserts, mapping -> next_mapping

let's make the name consistent in other methods as well.

cyx1231st reviewed (commented):
Code readability is important for repeated reviews and future maintainces, and is also critical to the correctness.

Review resources are valuable. Doing self-review before marking it ready can be very helpful expecially when the PR is large.

Making classes/methods self-contained and with clear signatures (names, comment, input, output) is crutial to reduce cognitive loads.

cyx1231st commented:
> Minor suggestions in/below fixed-kv-btree, mostly good.

If possible, let's split it out for merge.

cyx1231st commented:
> Offer 3 flexible options to deal with the left/right boundary:
> - remap boundary
> - delta-overwrite boundary
> - merge boundary if the extent is already pending

Let's make the logic clear about deciding the 3 possible options, the strategy can be improved later.

xxhdx1985126 commented:
@cyx1231st I've just addressed the comments and refactored ObjectDataHandler to improve its readability.

Specifically, overwrites in `ObjectDataHandler` are classified into two categories:
1. the range of the overwrite falls in a single lba mapping;
2. the range of the overwrite crosses multiple lba mappings.

For the first category, ObjectDataHandler processes the overwrites in
the following way:
1. if the mapping is a pending one (corresponds to a pending extent), merge the overwrite with the data of the pending extent;
2. otherwise, if the overwrite can, apply it with delta based overwrites;
3. otherwise, punch a hole in the mapping, insert a new extent with the data of the overwrite.

For the second category, the overwrite is processed as follows:
1. if the left boundary of the overwrite is inside an existing mapping, deal with the mapping in a way similar to the single-mapping   overwrites;
2. remove all lba mappings that are strictly within the range of the overwrite;
3. deal with the right boundary of the overwrite in the same way as the left boundary.

Please take a look again, thanks:-)

xxhdx1985126 commented:
The bottom up search capability in this PR is split into #64252

xxhdx1985126 commented:
jenkins test make check arm64

xxhdx1985126 commented:
Rebased:-

cyx1231st commented on src/crimson/os/seastore/object_data_handler.cc:
```diff
@@ -1083,309 +115,980 @@ ObjectDataHandler::write_ret ObjectDataHandler::prepare_data_reservation(
+  data_t &overwrite_data,
+  LBAMapping mapping)
+{
+  if (mapping.is_pending()) {
```
Why not reuse the central policy decision maker `get_edge_handle_policy()` ?

Otherwise it is not easy to maintain (adjust/unify) the decisions because they are not in the same place.

xxhdx1985126 replied:
Yeah, will do

xxhdx1985126 replied:
Sorry, this method should have been removed, seems I forgot it. Will correct it:-)

cyx1231st reviewed (commented):
Initial comments.

I'll be out until Monday, sorry if my response is delayed.

cyx1231st commented on src/crimson/os/seastore/object_data_handler.h:
```diff
@@ -255,6 +358,213 @@ class ObjectDataHandler {
+  };
+  enum edge_handle_policy_t : uint8_t {
+    DELTA_BASED_PUNCH,
+    MERGE_FROM_PENDING,
```
> For the initial-pending + overwriting-non-zero case, seems to me it is flexible to choose between the 2 merge operations.

So I think the 2 merge operations generally belong to the same decision. For initial-pending + overwriting-non-zero,  we can either do merge-inplace or merge-realloc (i.e. to merge the mappings if it is too fragmented).

xxhdx1985126 replied:
> > For the initial-pending + overwriting-non-zero case, seems to me it is flexible to choose between the 2 merge operations.
> 
> So I think the 2 merge operations generally belong to the same decision. For initial-pending + overwriting-non-zero, we can either do merge-inplace or merge-realloc (i.e. to merge the mappings if it is too fragmented).

Yeah, will try.

xxhdx1985126 commented on src/crimson/os/seastore/linked_tree_node.h:
```diff
@@ -884,54 +886,68 @@ class ParentNode {
+  // 1. The child extent is absent in cache
+  // 2. The child extent is (data) stable
+  //
+  // For reserved mappings, the return values are undefined.
```
Sorry, I'm not following. IIUC, the functionality of `_is_child_stable()` is not changed. Am I missing something?

xxhdx1985126 commented on src/crimson/os/seastore/object_data_handler.h:
```diff
@@ -78,6 +79,107 @@ class overwrite_buf_t {
+#ifndef NDEBUG
+  bool is_begin_in_mapping(const LBAMapping &mapping) const {
+    return unaligned_begin > mapping.get_key() &&
+      unaligned_begin < mapping.get_key() + mapping.get_length();
```
Um, I actually intended to check whether the begin/end was "strictly" in the mapping, maybe we can change the name to `is_begin/end_strictly_in_mapping()`, what do you think?

xxhdx1985126 commented on src/crimson/os/seastore/object_data_handler.h:
```diff
@@ -255,6 +358,213 @@ class ObjectDataHandler {
+    op_type_t op_type) const
+  {
+    //XXX: may need to adjust once object data partial write is available.
+    if (edge_mapping.is_pending()) {
```
?I don't understand this, if it is data stable, `edge_mapping.is_pending()` is false, and we'll do remap if it can't do delta based overwrite.

xxhdx1985126 commented on src/crimson/os/seastore/object_data_handler.h:
```diff
@@ -255,6 +358,213 @@ class ObjectDataHandler {
+      return edge_handle_policy_t::REMAP;
+    }
+
+    return edge_handle_policy_t::DELTA_BASED_PUNCH;
```
Yeah, but it seems that we may have to skip the `mapping.is_pending()` case, as `TransactionManager::remap_pin()` disallow remapping pending extents/mappings for now. We can add it in a followup PR, what do you think?

xxhdx1985126 commented on src/crimson/os/seastore/object_data_handler.h:
```diff
@@ -255,6 +358,213 @@ class ObjectDataHandler {
+
+    laddr_interval_set_t range;
+    range.insert(edge_mapping.get_key(), edge_mapping.get_length());
+    if (!range.contains(start, len)) {
```
Yeah, will change this into an assertion.

xxhdx1985126 commented on src/crimson/os/seastore/object_data_handler.h:
```diff
@@ -255,6 +358,213 @@ class ObjectDataHandler {
+      return edge_handle_policy_t::REMAP;
+    }
+
+    return edge_handle_policy_t::DELTA_BASED_PUNCH;
```
Yeah, added `op_type == op_type_t::ZERO` into the above `if` block

xxhdx1985126 commented on src/crimson/os/seastore/object_data_handler.cc:
```diff
@@ -1083,309 +115,980 @@ ObjectDataHandler::write_ret ObjectDataHandler::prepare_data_reservation(
+      assert(overwrite_range.aligned_end <
+	    edge_mapping.get_key() + edge_mapping.get_length());
+      if (overwrite_range.aligned_end <
+		edge_mapping.get_key() + edge_mapping.get_length()) {
```
Sorry, the assert above shouldn't exist.

xxhdx1985126 commented on src/crimson/os/seastore/object_data_handler.cc:
```diff
@@ -1083,309 +115,980 @@ ObjectDataHandler::write_ret ObjectDataHandler::prepare_data_reservation(
+
+// read the padding edge data into data.headbl/data.tailbl, note that
+// the method doesn't expand the overwrite range, as the aligned boundaries
+// are not affected, expands only happens in the merge_pending_edge method.
```
?I don't understand this, there's no need to expand/shrink `overwrite_range` here.

xxhdx1985126 commented on src/crimson/os/seastore/object_data_handler.cc:
```diff
@@ -1083,309 +115,980 @@ ObjectDataHandler::write_ret ObjectDataHandler::prepare_data_reservation(
-  if (bl.has_value()) {
-    assert(bl->length() == len);
+  if (right_mapping.is_end() ||
+      overwrite_range.aligned_end <= right_mapping.get_key()) {
```
Um, not really, if the last mapping in the range of the hole end exactly at `overwrite_range.aligned_end`, `TransactionManager::remove_mappings_in_range()` will return the next mapping, which is `key <= overwrite end`.

I think the reason that `right_mapping` is different from `left_mapping` at this point is that `left_mapping` is retrieved through `TM::get_containing_pin()` which is supposed to return the mapping that contains the laddr, while `right_mapping` is always the next mapping of the last mapping in the hole range.

xxhdx1985126 replied:
But maybe we can move this `if` block out of the `punch_right_mapping()` and into `punch_multi_mapping_hole`

xxhdx1985126 commented:
@cyx1231st I've just addressed all the comments except for the ones that are still under discussion. Please take a look, again, thanks:-)

xxhdx1985126 commented:
Just refined `LBAManager::ref_update_result_t`.

Previously, if `LBAManager::remove_mapping` caused the intermediate direct mapping to be removed, `LBAManager::ref_update_result_t` would contain the info of the removed direct mapping, instead of the removed indirect mapping.

Now, it contains the info of both mappings.

xxhdx1985126 commented:
> The following tests FAILED:
	209 - unittest_bluefs (SEGFAULT)
	210 - unittest_bluefs_ex (Failed)
	212 - unittest_bdev (SEGFAULT)

xxhdx1985126 commented:
jenkins test make check

xxhdx1985126 commented:
jenkins test api

xxhdx1985126 commented:
https://pulpito.ceph.com/xuxuehan-2025-08-03_17:05:49-crimson-rados-wip-seastore-transaction-manager-iterator-ci-test-distro-default-smithi/

xxhdx1985126 commented:
https://pulpito.ceph.com/xuxuehan-2025-08-05_11:19:55-crimson-rados-wip-seastore-transaction-manager-iterator-ci-test-distro-crimson-debug-smithi/

Seems that the only failure was not relevant: https://pulpito.ceph.com/xuxuehan-2025-08-05_11:19:55-crimson-rados-wip-seastore-transaction-manager-iterator-ci-test-distro-crimson-debug-smithi/8425179/

I added three more commits to fix the bugs in this PR:
1. crimson/os/seastore/lba_mapping: treat indirect mappings as valid non-zero-reserved mappings
2. crimson/os/seastore/lba_mapping: treat indirect mappings as valid non-clone ones
3. crimson/os/seastore/transaction_manager: complete indirect mappings before reading

Now the tests seems good, @athanatos could you please take a final look at this PR? Thanks:-)

xxhdx1985126 commented:
>[ https://pulpito.ceph.com/xuxuehan-2025-08-05_11:19:55-crimson-rados-wip-seastore-transaction-manager-iterator-ci-test-distro-crimson-debug-smithi/](https://wx.qq.com/cgi-bin/mmwebwx-bin/webwxcheckurl?requrl=https%3A%2F%2Fpulpito.ceph.com%2Fxuxuehan-2025-08-05_11%3A19%3A55-crimson-rados-wip-seastore-transaction-manager-iterator-ci-test-distro-crimson-debug-smithi%2F&skey=%40crypt_b56163cf_ab56976b8ba466bb4ce459e5437b999d&deviceid=e714322076423869&pass_ticket=KDkOlqYLENWf5zM696Bzq1wIc3fMxX6yob0%252FSs3IXIdm3qw7iOpuUCWB2oKWYEMUM%252Fzn0g3Q1gbyoR1Ki089CA%253D%253D&opcode=2&scene=1&username=@0209bb6feb515aed295db339ff766275)
> 
> ~Seems that the only failure was not relevant~ looking into:[ https://pulpito.ceph.com/xuxuehan-2025-08-05_11:19:55-crimson-rados-wip-seastore-transaction-manager-iterator-ci-test-distro-crimson-debug-smithi/8425179/](https://wx.qq.com/cgi-bin/mmwebwx-bin/webwxcheckurl?requrl=https%3A%2F%2Fpulpito.ceph.com%2Fxuxuehan-2025-08-05_11%3A19%3A55-crimson-rados-wip-seastore-transaction-manager-iterator-ci-test-distro-crimson-debug-smithi%2F8425179%2F&skey=%40crypt_b56163cf_ab56976b8ba466bb4ce459e5437b999d&deviceid=e714322076423869&pass_ticket=KDkOlqYLENWf5zM696Bzq1wIc3fMxX6yob0%252FSs3IXIdm3qw7iOpuUCWB2oKWYEMUM%252Fzn0g3Q1gbyoR1Ki089CA%253D%253D&opcode=2&scene=1&username=@0209bb6feb515aed295db339ff766275)

I tried to reproduce this error, however it hasn't popped up during the last several teuthology runs. I submitted an issue for it:[ https://tracker.ceph.com/issues/72459](https://wx.qq.com/cgi-bin/mmwebwx-bin/webwxcheckurl?requrl=https%3A%2F%2Ftracker.ceph.com%2Fissues%2F72459&skey=%40crypt_b56163cf_ab56976b8ba466bb4ce459e5437b999d&deviceid=e714322076423869&pass_ticket=KDkOlqYLENWf5zM696Bzq1wIc3fMxX6yob0%252FSs3IXIdm3qw7iOpuUCWB2oKWYEMUM%252Fzn0g3Q1gbyoR1Ki089CA%253D%253D&opcode=2&scene=1&username=@0209bb6feb515aed295db339ff766275)

According to its failure "open_collection failed", I think it's not relevant to this PR and this PR should be good to go:-)

athanatos commented:
jenkins test make check arm64

Matan-B commented:
https://pulpito.ceph.com/xuxuehan-2025-08-05_11:19:55-crimson-rados-wip-seastore-transaction-manager-iterator-ci-test-distro-crimson-debug-smithi/
